File: web-app/backend/app/schemas/installer.py
"""Pydantic models for /installer-requests (Execution Marketplace Phase 0)."""
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class InstallerRequestCreate(BaseModel):
    """Body for POST /installer-requests (authenticated customer)."""

    fabric_building_id: str = Field(description="The building the measure is for.")
    action_type: str | None = Field(
        default=None, description="Recommended measure token, e.g. INSTALL_HEAT_PUMP."
    )
    measure_label: str | None = Field(
        default=None, description="Human label of the measure (for the founder's queue)."
    )
    source: str | None = Field(
        default=None, description="Where it was raised: 'actions' | 'residential'."
    )


class InstallerRequestAck(BaseModel):
    """The confirmation returned to the customer."""

    id: UUID
    status: str
    message: str


class InstallerRequestRow(BaseModel):
    """One request in the founder's broker queue."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    organization_id: UUID
    fabric_building_id: str
    building_name: str | None = None
    action_type: str | None = None
    measure_label: str | None = None
    note: str | None = None
    source: str | None = None
    status: str
    created_at: datetime


class InstallerRequestList(BaseModel):
    """Payload for GET /installer-requests/admin."""

    requests: list[InstallerRequestRow] = Field(default_factory=list)


class InstallerStatusUpdate(BaseModel):
    """Body for PATCH /installer-requests/admin/{id}."""

    status: str = Field(description="requested | contacted | quoted | closed")

File: web-app/frontend/lib/api/installer.ts
/**
 * Installer-request API helpers (Execution Marketplace Phase 0).
 *
 * Customer create goes through /api/installer-requests (the button posts it
 * directly). Admin reads are server-side (founder queue); the status PATCH goes
 * through the /api/installer-requests/admin proxy. Types mirror backend
 * app/schemas/installer.py.
 */

export type AdminInstallerRequestRow = {
  id: string
  organization_id: string
  fabric_building_id: string
  building_name: string | null
  action_type: string | null
  measure_label: string | null
  note: string | null
  source: string | null
  status: string
  created_at: string
}

export type AdminInstallerRequestsResponse = {
  requests: AdminInstallerRequestRow[]
}

type Result<T> = { ok: true; data: T } | { ok: false; error: string }
type ServerResult<T> =
  | { ok: true; data: T }
  | { ok: false; error: string; status?: number }

export async function fetchAdminInstallerRequestsServer(
  token: string
): Promise<ServerResult<AdminInstallerRequestsResponse>> {
  const base = process.env.BACKEND_URL
  if (!base) return { ok: false, error: "Server misconfigured" }
  try {
    const res = await fetch(`${base}/installer-requests/admin`, {
      headers: { Authorization: `Bearer ${token}`, Accept: "application/json" },
      cache: "no-store",
    })
    if (!res.ok) {
      return { ok: false, error: `Backend ${res.status}`, status: res.status }
    }
    return { ok: true, data: (await res.json()) as AdminInstallerRequestsResponse }
  } catch (err) {
    console.error("[fetchAdminInstallerRequestsServer] Network error:", err)
    return { ok: false, error: "Network error" }
  }
}

export async function updateInstallerRequestStatus(
  requestId: string,
  status: "requested" | "contacted" | "quoted" | "closed"
): Promise<Result<unknown>> {
  try {
    const res = await fetch(
      `/api/installer-requests/admin/${encodeURIComponent(requestId)}`,
      {
        method: "PATCH",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ status }),
      }
    )
    if (!res.ok) {
      const body = await res.json().catch(() => ({}))
      return {
        ok: false,
        error: (body && (body.detail as string)) || `Failed (${res.status})`,
      }
    }
    return { ok: true, data: await res.json().catch(() => ({})) }
  } catch {
    return { ok: false, error: "Network error" }
  }
}
